code/parse/sexp.cpp

```diff
@@ -23298,6 +23567,706 @@ void sexp_set_motion_debris(int node)
+			return SEXP_TRUE;
+		}
+	} else if (Sexp_containers[index].is_list()) {
+		if (Sexp_containers[index].list_data.empty()) {
+			return SEXP_TRUE;
+		}
+	} else {
+		int type = Sexp_containers[index].type; 
+		Error(LOCATION, "Unknown container type. Container %s is of type %d and this is not a valid type", container_name, type);
+	}
+
+	return SEXP_FALSE;
+}
+
+int sexp_get_container_size(int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name);
+		return 0;
+	}
+
+	Assert(index < (int)Sexp_containers.size());
+
+	const auto &container = Sexp_containers[index];
+	if (container.is_map()) {
+		return (int)container.map_data.size();
+	} else if (container.is_list()) {
+		return (int)container.list_data.size();
+	}
+
+	Error(LOCATION, "Invalid container type: %d", container.type);
+	return 0;
+}
+
+void sexp_clear_container (int node)
+{
+	const char *container_name = CTEXT(node);
+	const int index = get_sexp_container_index(container_name);
+
+	if (index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", container_name); 
+		return;
+	}
+
+	Assert(index < (int)Sexp_containers.size());
+
+	auto &container = Sexp_containers[index];
+	if (container.is_map()) {
+		container.map_data.clear();
+	} else if (container.is_list()) {
+		container.list_data.clear();
+	} else {
+		Error(LOCATION, "Unknown container type. Container %s is of type %d and this is not a valid type", container_name, container.type);
+	}
+}
+
+void sexp_get_map_keys(int node)
+{
+	bool replace_contents = true;
+
+	const char *map_container_name = CTEXT(node);
+	const int map_index = get_sexp_container_index(map_container_name);
+
+	if (map_index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", map_container_name); 
+		return;
+	}
+
+	node = CDR(node); 
+	const char *list_container_name = CTEXT(node);
+
+	const int list_index = get_sexp_container_index(list_container_name);
+
+	if (list_index < 0) {
+		Warning(LOCATION, "Container %s does not exist.", list_container_name); 
+		return;
+	}
+
+	// both containers must be of the correct type. 
+	if (!Sexp_containers[map_index].is_map() || !Sexp_containers[list_index].is_list()) {
+		Warning(LOCATION, "Containers must be of the correct types to be used in get-map-keys SEXP."); 
+		return;
+	}
+
+	node = CDR(node); 
+	if (node != -1) {
+		replace_contents = is_sexp_true(node); 
+	}
+
+	if (replace_contents) {
+		Sexp_containers[list_index].list_data.clear(); 
+	}
+
+	for (const auto &kv_pair : Sexp_containers[map_index].map_data) {
+		Sexp_containers[list_index].list_data.emplace_back(kv_pair.first);
+	}
+}
+
+int get_sexp_container_index_generic(const char *name,
+	const std::function<bool(const char*, const char*)>& match_func)
+{
+	const char initial_lower = SCP_tolower(name[0]);
+
+	if (Container_indices_by_initial.count(initial_lower) == 0) {
```

Comment:
The usefulness of `Container_indices_by_initial` is debatable, although it can make verifying that a container name is *invalid* fast. If the name to check starts with a letter that no actual container's name starts with, then we can return -1 right away (i.e., in the line right below this one).